### DISCUSSION TITLE
Move away from assert

### DISCUSSION THREAD

**Discussion by jonas-eschle** [pydantic/pydantic#4575] [Pydantic V2]
Hi there, let me know if I should make improvement proposals somewhere else.

Getting familiar with pydantic, I've noticed a few things, so posting theme here.

Across pydantic, assert is used to check if a value is fine. This is wrong! An `assert` is a sanity check inside your code in order to detect *internal* errors.
If arguments to a function are wrong, they should never be tested with assert but raise exceptions if they don't comply. An assert _may_ be used to check that you could convert a value succesfully (i.e. an `assert isinstance(value, str)`, if you rely on the fact that you could successfully convert value to a string, it just checks if there is no hole in the logic).

Anything else, especially in validation of values, should always be an exception.

I suggest to change this in the code and docs for V2.

**Comment by andrew222651** [comment]
Let me understand your point. Some people use asserts for debugging or type annotations and want them disabled in production for performance. Some people use asserts for [contracts](https://crosshair.readthedocs.io/en/latest/kinds_of_contracts.html#assert-based-contracts) and may or may not want them disabled in production. If the two styles are mixed in the same Python instance there's a conflict. And/or people should raise more specific exceptions instead of just `AssertionError`. Is that the reasoning or something else?

**Comment by samuelcolvin** [comment]
> Across pydantic, assert is used to check if a value is fine.

This is incorrect. Asserts not used in pydantic, rather `AssertionError` (along with `ValueError`, and `TypeError` in v1) are caught and treated as validation errors.

Pydantic built in validators do not use assert statements and will therefore work even if you disable debug.

`AssertionError` will continue to be respected in v2 for those who want to use it.

**Reply by jonas-eschle** [reply]
I partially disagree: the documentation mentions at several points - and uses in the examples ([see here](https://github.com/pydantic/pydantic/blob/e34ff9255f281eaccac10d0436c082626d9146c3/docs/examples/validators_subclass_each_item.py#L12)) - to use the "assert" statement.

The problem is, as mentioned also above, that this is not what asserts are made for and can easily be a bug: what if code raises an *actual* assert (because a library does an internal check that isn't caught?). pydantic will silently assume that the assertion error was triggered by an assert in a validator.

Why not decouple the business logic (as @amacfie also mentioned): create a _specific_ exception (pydantic.ValidationError) (for compatibility, we can also make this a subclass of other, currently used Exceptions, until the API actually breaks to give users time to adjust) that users should raise. No other library will ever raise this error and we solve:
a) unintentionally removed asserts in user code (which is promoted in examples)
b) unintentionally catching a ValueError for example that wasn't meant to be caught (say because a validator creates an actual ValueError)
c) catch no internal errors (asserts) of other libraries.
d) be explicit! A ValidationError is pretty expressive. And it would even allow for a custom error message that pydantic can provide or print options (whether to print "value" or this or that...)

**Comment by samuelcolvin** [comment]
We have a custom error for V2 to provide more information, but we also allow `ValueError` and `AssertionError`.

I think what you're missing is that this isn't in all your code, or even a significant part of it. It's **only** inside functional validators.

Also it's not like this code silently passes, rather that a validation error is raised.

@PrettyWood @hramezani @tiangolo what do you think?

**Reply by jonas-eschle** [reply]
> We have a custom error for V2 to provide more information

That sounds good!

> Also it's not like this code silently passes 

That's maybe currently the logic, but is that future proof? What about a trial-and-error type, i.e. to resolve unions? Try all possible ones and see which works? These kind of mechanics could introduce bugs.

> I think what you're missing is that this isn't in all your code

No, I think you're missing a point here. It **is** in all of my code. Because `assert` is a Python function that can help me check the validity of *my own code*. It's not a `pydantic` exclusive object (whereas the custom error is and your argument fully applies to that).

But the most striking problem to me seems the optimization: I strongly believe that the optimization of the Python files should not be directly coupled to whether *some* validations work or not. A switch to turn of validations, why not? Only some, why not? But keep in mind, the assert is an error that is only raised if `__debug__` is true. But this should not has to do with how the Python file was compiled. The optimization flag of code should **never** change the logic of the code (that's the whole point of optimization).

The docs describe them as [Assert statements are a convenient way to insert debugging assertions into a program](https://docs.python.org/3/reference/simple_stmts.html#the-assert-statement) and mention that it's probably better to explicitly raise an `AssertionError` if really needed.

But then again, why? `exception.AnyError` is as complicated as `pydantic.ValidationError`.

(as a sidenote, we have done quite the same "mistake" first in our implementation for the logical flow, using the "NotImplementedError" to signal that a fallback should be used. IMHO, and after years of thinking and discussing, there are just downsides and your own exception is just the "only" right solution: It's verbose, it does **exactly** what you want and nothing else (nobody else ever catches it, nobody else ever throws it) and it's always there)

And glad to hear about all the V2 improvements, I very much agree that the way forward is to just... forget V1 and focus on V2 (as cumbersome as this may be as a user), definitely do that! That's why I am trying to bring suggestions so that V2 is right, because I guess now it still can be changed.
